tests/src/main/java/net/neoforged/neoforge/debug/transfer/VanillaHandlersTests.java

```diff
@@ -0,0 +1,566 @@
+//        Inventory inventory = HopperBlockEntity.getInventoryAt(context.getWorld(), context.getAbsolutePos(chestPos));
+//        context.assertTrue(inventory != null, Text.literal("Inventory must not be null"));
+//
+//        if (!(storage instanceof SlottedStorage<ItemResource> slottedStorage)) {
+//            throw context.assertionException("Double chest storage must be a SlottedStorage");
+//        }
+//
+//        for (int i = 0; i < inventory.size(); ++i) {
+//            ItemStack stack = inventory.getStack(i);
+//            ItemResource variant = ItemResource.of(stack.getItem());
+//            context.assertTrue(variant.matches(stack), Text.literal("Item variant in slot " + i + " must match stack"));
+//            long expectedCount = stack.getCount();
+//            long actualCount = slottedStorage.getSlot(i).getAmount();
+//            context.assertTrue(expectedCount == actualCount, Text.literal("Slot " + i + " should have " + expectedCount + " items, but has " + actualCount));
+//        }
+//
+//        // Check that an update is queued for every single comparator
+//        MutableInt comparatorCount = new MutableInt();
+//
+//        context.forEachRelativePos(relativePos -> {
+//            if (context.getBlockState(relativePos).getBlock() != Blocks.COMPARATOR) {
+//                return;
+//            }
+//
+//            comparatorCount.increment();
+//
+//            if (!context.getWorld().getBlockTickScheduler().isQueued(context.getAbsolutePos(relativePos), Blocks.COMPARATOR)) {
+//                throw context.assertionException("Comparator at " + relativePos + " should have an update scheduled");
+//            }
+//        });
+//
+//        context.assertTrue(comparatorCount.intValue() == 6, Text.literal("Expected exactly 6 comparators"));
+//
+//        helper.succeed();
+//    }
+
+    /**
+     * Regression test for <a href="https://github.com/FabricMC/fabric/issues/3017">composters not always incrementing their level on the first insert</a>.
+     */
+    @GameTest
+    @EmptyTemplate
+    @TestHolder
+    public static void testComposterFirstInsert(ExtendedGameTestHelper helper) {
+        BlockPos pos = new BlockPos(0, 1, 0);
+
+        ItemResource carrot = ItemResource.of(Items.CARROT);
+
+        for (int i = 0; i < 200; ++i) { // Run many times as this can be random.
+            helper.setBlock(pos, Blocks.COMPOSTER.defaultBlockState());
+            var wrapper = ComposterWrapper.get(helper.getLevel(), helper.absolutePos(pos), Direction.UP);
+
+            try (Transaction tx = Transaction.open(null)) {
+                if (wrapper.insert(carrot, 1, tx) != 1) {
+                    helper.fail(Component.literal("Carrot should have been inserted"), pos);
+                }
+
+                tx.commit();
+            }
+
+            helper.assertBlockState(pos, state -> state.getValue(ComposterBlock.LEVEL) == 1, (s) -> Component.literal("Composter should have level 1"));
```

Comment:
Personally I am not a fan of the parenthesis around a single variable lambda, but regardless of that, can we make use of the parameter and specify in the error what the actual level currently is?